### PR TITLE
Batch snapshot create tasks

### PR DESCRIPTION
**Pull request by original-brownbear**
Batch most snapshot related cluster state update tasks into a single executor by reusing the already batched executor for shard state updates.
This required a rather straightforward refactoring of the consistent cluster state update into the `SnapshotsService` but is otherwise rather straight forward since I did not change the tasks themselves in any way, just made them get applied in the executor.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
I think this is ok but it'd help me to see the refactoring as a separate PR from the behaviour change. Could you split them up?

**Comment by original-brownbear** [comment]
Exctracted the refactoring of the consistent state update into https://github.com/elastic/elasticsearch/pull/94020 now, that should make this a lot easier to read.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -3576,4 +3406,184 @@ private record ShardSnapshotMetaDeleteResult(
...
+                    Metadata.builder(currentState.getMetadata())
+                        .putCustom(
+                            RepositoriesMetadata.TYPE,
+                            currentState.metadata()
+                                .<RepositoriesMetadata>custom(RepositoriesMetadata.TYPE)
+                                .withUpdatedGeneration(repoName, safeGeneration, newGen)
+                        )
+                        .build()
+                )
+                .build();
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:3475]: NIT: could be simplified using `ClusterState#copyAndUpdateMetadata`

File: server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryRestoreTests.java [hunk 2]
```diff
@@ -144,6 +144,7 @@ public void testRestoreSnapshotWithExistingFiles() throws IOException {
         }
     }
 
+    @AwaitsFix(bugUrl = "foo")
```
Comment (idegtiarenko) [inline server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryRestoreTests.java:147]: Is it still failing?

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-1

File: server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryRestoreTests.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryRestoreTests.java]: Nah my bad :)

**Review by idegtiarenko — APPROVED** [review]
:+1: from my side, but I would like somebody else with better `ClusterStateTaskExecutor` knowledge to approve as well

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I would rather we didn't use mutable fields on the task to pass information from `execute()` to `clusterStateProcessed()`, and I'm also suspicious about passing `newState` into `clusterStateProcessed()` because this will be the _final_ state at the end of the batch, not the state after processing each item, and that tends to be a source of bugs.

**Comment by original-brownbear** [comment]
> I'm also suspicious about passing newState into clusterStateProcessed() because this will be the final state at the end of the batch, not the state after processing each item, and that tends to be a source of bugs.

This is what we want. The implicit assumption in all of this code at the moment is that the callback sees the final version of the state and triggers the next action based on it.

 > I would rather we didn't use mutable fields on the task to pass information from execute() to clusterStateProcessed()

That's probably something that I can fix ... let me try :)

**Comment by original-brownbear** [comment]
Ok nevermind.

> I would rather we didn't use mutable fields on the task to pass information from execute() to clusterStateProcessed()

This isn't so trivial to fix. For cloning in particular we'd have to make some bigger changes here from how things work today. Maybe just do that in a follow-up that removes the use of `ClusterState` returns from these tasks in general like the TODO I added suggests?
I really need different tasks to work without the mutable fields :)

**Review by DaveCTurner — COMMENTED** [review]
I'm not sure I see the particular difficulty with the clone operation, we just have to pass the `newEntry` down the line don't we?

Also I didn't see any places where we really need to see the final cluster state, at least not if we fix the information flow. We're pulling out the latest `SnapshotsInProgress` in some places, but I reckon this makes more sense at the batch level than in a handler on each individual task. The other things we use are stuff like `.metadata()` and `.nodes()` that won't be changing during these tasks.

**Comment by original-brownbear** [comment]
> Also I didn't see any places where we really need to see the final cluster state, at least not if we fix the information flow.

I really fail to see the value of changing the mechanics here tbh. We do actions in the callback based on the final cluster state. If we start doing actions based on some intermediary state that resulted from just the single update, things just become incredibly hard to reason about.

> I'm not sure I see the particular difficulty with the clone operation, we just have to pass the newEntry down the line don't we?

Yea it's not hard but then I can't just reuse the tasks as is and must work this into the executor in some form by returning the entry from the task and setting it as the task result later. I can do that here, but that's gonna be more changes. And we do have a couple of these spots in two cloning updates, in the repo generation setting etc. making this a bit of a bigger PR which I tried to avoid.

**Comment by DaveCTurner** [comment]
> If we start doing actions based on some intermediary state that resulted from just the single update, things just become incredibly hard to reason about.

Sorry, I wasn't clear. I'm ok with deciding what to do based on the final state, but we should make those decisions in the executor when it reaches the end of the batch, rather than doing it on a task-by-task basis as proposed here.

I think we can/should make these changes to each task separately first, and then this change to add batching will be trivial and obviously correct.

**Comment by original-brownbear** [comment]
> I think we can/should make these changes to each task separately first, and then this change to add batching will be trivial and obviously correct.

Hmm and then add a new separate executor per task but not batch to begin with? This sounds like it'll turn into quite the project :) Let me try.

> Sorry, I wasn't clear. I'm ok with deciding what to do based on the final state, but we should make those decisions in the executor when it reaches the end of the batch, rather than doing it on a task-by-task basis as proposed here.

Seems to me that's almost just more code with the way things work currently. Yes if we refactored things nicely and wouldn't be generating a series of states it may be nice to just handle the callback logic in the executor. But as things are now and without bigger changes, it's really just an excercise in adding more explicit fields to the tasks so that the executor can then run the code we have in the callback at present?

**Comment by original-brownbear** [comment]
Then again, seems like this PR just got nuked by the changes to task submitting to a good degree :( -> closing here for now, I'll see if I can find the time to revisit this step by step.

**Comment by DaveCTurner** [comment]
Ah yeah sorry about that. It shouldn't be too much of a big deal to fix it up for those changes - I can do the master merge if you'd like.

**Comment by DaveCTurner** [comment]
Ah ok I see yes it's a bit of a pain to share the queue across the repositories service and the snapshots service. I can hack something in I reckon but maybe this needs a bit more of a rethink.

**Comment by original-brownbear** [comment]
Yea this was though as a quick win but yea this got too complicated now ... I'll try something smaller for the time being.

**Pull request closed without merging**